File: yidam/editors/vscode/src/graph.ts
/**
 * Link navigation, as data.
 *
 * # What this file is allowed to know
 *
 * A corpus edge is a filesystem-relative path inside YAML, resolved as `dir.join(target)`
 * against the instance's own directory and normalized. That rule belongs to `dangling_edge`
 * and `orphan_in`, and **the CLI applies it** — `yidam graph --format json` reports every
 * edge already resolved, with `exists` answered by the same test the gate uses. Nothing here
 * decides whether an edge is broken.
 *
 * What is left is genuinely navigation: reading the scalar under the cursor out of one line
 * of text, joining a path so a `Go to Definition` works in a buffer that has not been saved
 * since the report was taken, and ranking a completion list. When any of it is wrong you
 * fail to jump. The verdict still comes from `lint`.
 *
 * # The ontology is a guide, not a closed list
 *
 * Measured against a live derived repository at 90 nodes and 299 edges: **17 of the
 * (class, relationship) pairs actually in use are not declared as `out` edges on their
 * class**, and one of them — `instance-of`, the edge every node carries to its own
 * `.ont.yml` — is used by every class and declared by none. Nothing lints relationships
 * against the ontology, so a completion list restricted to declared edges would be stricter
 * than any rule in the system and would omit the corpus's single most-used relationship.
 *
 * So: declared edges first, relationships already in use beside them, and the reason each
 * one is offered in its detail text. This is the opposite of the commit vocabulary, which
 * *is* closed and *is* gated — and the difference is why one offers a squiggle and the
 * other does not.
 *
 * No `vscode` import.
 */

import type { Envelope } from './reports.ts'

export interface GraphLink {
  target: string
  relationship: string
  /** Corpus-relative, resolved by the CLI. Empty when it lands outside the corpus. */
  resolved: string
  exists: boolean
}

export interface GraphNode {
  /** Corpus-relative — this is a node's identity throughout. */
  node: string
  class: string
  label: string
  description: string
  links: GraphLink[]
}

export interface OntProperty {
  name: string
  type: string
  description: string
}

export interface OntEdge {
  relationship: string
  /** A class name, not a path. */
  target: string
  direction: string
  description: string
}

export interface GraphClass {
  class: string
  label: string
  description: string
  properties: OntProperty[]
  edges: OntEdge[]
}

export interface GraphReport extends Envelope {
  corpus_dir: string
  nodes: GraphNode[]
  classes: GraphClass[]
}

// ── reading the line under the cursor ────────────────────────────────────────

export interface Scalar {
  value: string
  /** 0-based character offsets of the value, so a range can be drawn on it. */
  start: number
  end: number
}

/**
 * The value of `key:` on this line, if the cursor is inside it.
 *
 * One line of text, not a YAML parse. Two reasons, and the second is the load-bearing one:
 * the buffer may differ from what the report was taken over, so a lookup by document
 * position has to read the document; and parsing corpus YAML in TypeScript would be a
 * second implementation of `parse_node`, which is one of the six parity functions.
 */
export function scalarAt(line: string, character: number, key: string): Scalar | null {
  const m = new RegExp(`^(\\s*(?:-\\s*)?${key}:\\s*)(.*)$`).exec(line)
  if (!m) return null
  const start = m[1].length
  let value = m[2]
  // Trailing comment, then trailing space. A quoted value keeps its quotes out of the range
  // so the path is what gets joined.
  const hash = value.indexOf(' #')
  if (hash !== -1) value = value.slice(0, hash)
  value = value.replace(/\s+$/, '')
  if (value.length === 0) return null
  let offset = 0
  if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
    if (value.length < 2) return null
    offset = 1
    value = value.slice(1, -1)
  }
  const from = start + offset
  const to = from + value.length
  // Inclusive of both ends: clicking just past the last character is still the value.
  if (character < from || character > to) return null
  return { value, start: from, end: to }
}

/**
 * `concept/tailwater.yml` + `../gauge/g.yml` → `gauge/g.yml`.
 *
 * Corpus-relative in, corpus-relative out — the same identity the report uses. Mirrors the
 * CLI's `normalize(dir.join(target))`; a `..` that climbs above the corpus root yields the
 * empty string, which is what the CLI reports for the same case.
 */
export function resolveFrom(nodeId: string, target: string): string {
  const dir = nodeId.slice(0, nodeId.lastIndexOf('/') + 1)
  const parts: string[] = []
  for (const part of `${dir}${target}`.split('/')) {
    if (part === '' || part === '.') continue
    if (part === '..') {
      if (parts.length === 0) return ''
      parts.pop()
      continue
    }
    parts.push(part)
  }
  return parts.join('/')
}

/** `gauge/g.yml` seen from `concept/tailwater.yml` → `../gauge/g.yml`. */
export function relativeFrom(nodeId: string, targetId: string): string {
  const from = nodeId.split('/').slice(0, -1)
  const to = targetId.split('/')
  let i = 0
  while (i < from.length && i < to.length - 1 && from[i] === to[i]) i += 1
  const up = '../'.repeat(from.length - i)
  const down = to.slice(i).join('/')
  return `${up}${down}` || down
}

/**
 * The line holding `target: <raw>`, 0-based, or 0 when it is not found.
 *
 * A text scan, and only ever used to *place a cursor*. Falling back to line 0 rather than to
 * "no reference" is deliberate: the edge is in the report, so it is in the file, and pointing
 * at the top of the right file beats dropping a reference the graph says exists.
 */
export function lineOfTarget(text: string, rawTarget: string): number {
  const lines = text.split('\n')
  for (let i = 0; i < lines.length; i += 1) {
    const found = scalarAt(lines[i], -1, 'target')
    if (found === null) {
      // `scalarAt` gates on the cursor; re-read without one.
      const m = /^\s*(?:-\s*)?target:\s*(.*)$/.exec(lines[i])
      if (m && m[1].replace(/\s+$/, '').replace(/^["']|["']$/g, '') === rawTarget) return i
      continue
    }
    if (found.value === rawTarget) return i
  }
  return 0
}

// ── lookups ──────────────────────────────────────────────────────────────────

export function nodeById(graph: GraphReport, id: string): GraphNode | undefined {
  return graph.nodes.find((n) => n.node === id)
}

export interface Reference {
  /** The node holding the edge. */
  from: string
  relationship: string
  /** The raw target text, so the line can be found in the file. */
  target: string
}

/**
 * Every inbound edge to a node.
 *
 * Nothing surfaced reverse traversal for corpus nodes: `used-by` covers catalog entries
 * only, and `orphan-in` reports the *absence* of inbound edges without ever naming the
 * present ones.
 */
export function referencesTo(graph: GraphReport, id: string): Reference[] {
  const out: Reference[] = []
  for (const n of graph.nodes) {
    for (const l of n.links) {
      if (l.resolved === id) out.push({ from: n.node, relationship: l.relationship, target: l.target })
    }
  }
  return out
}

/** The hover for a link: what is on the other end, without leaving the node. */
export function hoverFor(graph: GraphReport, id: string): string | null {
  const node = nodeById(graph, id)
  if (!node) {
    // An `.ont.yml` is a legitimate target — every instance carries an `instance-of` edge
    // to one — and it is not a node, so say what it is rather than nothing.
    if (id.endsWith('.ont.yml')) {
      const cls = graph.classes.find((c) => `${c.class}.ont.yml` === id)
      if (cls) return `**${cls.label || cls.class}** — \`${id}\`\n\n${cls.description}`
    }
    return null
  }
  const inbound = referencesTo(graph, id).length
  const meta = `\`${node.class}\` · ${node.links.length} out · ${inbound} in`
  return `**${node.label || node.node}** — ${meta}\n\n${node.description}`
}

// ── completion ───────────────────────────────────────────────────────────────

export interface Candidate {
  label: string
  detail: string
  documentation: string
  /** 0 sorts first. Declared beats observed beats everything. */
  rank: number
}

function sortKey(c: Candidate, i: number): string {
  return `${c.rank}${String(i).padStart(4, '0')}`
}

export function sorted(cs: Candidate[]): (Candidate & { sortText: string })[] {
  return cs
    .slice()
    .sort((a, b) => a.rank - b.rank || a.label.localeCompare(b.label))
    .map((c, i) => ({ ...c, sortText: sortKey(c, i) }))
}

/**
 * Out-edges the class declares, then relationships its instances already use.
 *
 * **One relationship may be declared against several classes.** A real ontology has
 * `maneuver -[operates-on]->` legislation, ballot-measure *and* election as three separate
 * declarations; three classes there do it. They are one offer with three destinations, and
 * treating each declaration as its own candidate put `concerns` in the list twice.
 */
export function declaredOut(graph: GraphReport, cls: string, relationship?: string): OntEdge[] {
  return (graph.classes.find((c) => c.class === cls)?.edges ?? []).filter(
    (e) => e.direction !== 'in' && (relationship === undefined || e.relationship === relationship),
  )
}

export function relationshipCandidates(graph: GraphReport, cls: string): Candidate[] {
  const byRelationship = new Map<string, OntEdge[]>()
  for (const e of declaredOut(graph, cls)) {
    byRelationship.set(e.relationship, [...(byRelationship.get(e.relationship) ?? []), e])
  }
  const out: Candidate[] = [...byRelationship.entries()].map(([relationship, edges]) => ({
    label: relationship,
    detail: edges.some((e) => e.target)
      ? `→ ${edges.map((e) => e.target).filter(Boolean).join(' | ')}`
      : 'declared',
    documentation: edges
      .map((e) => (edges.length > 1 && e.target ? `**${e.target}** — ${e.description}` : e.description))
      .filter(Boolean)
      .join('\n\n'),
    rank: 0,
  }))

  const named = new Set(out.map((c) => c.label))
  const inUse = new Map<string, number>()
  for (const n of graph.nodes) {
    if (n.class !== cls) continue
    for (const l of n.links) {
      if (!l.relationship || named.has(l.relationship)) continue
      inUse.set(l.relationship, (inUse.get(l.relationship) ?? 0) + 1)
    }
  }
  for (const [relationship, count] of inUse) {
    out.push({
      label: relationship,
      detail: `in use — ${count}×`,
      documentation:
        `Used by ${count} \`${cls}\` node(s) and not declared in \`${cls}.ont.yml\`. ` +
        'Nothing gates relationships against the ontology, so this is offered rather than ' +
        'hidden — but it is worth knowing the ontology does not describe it.',
      rank: 1,
    })
  }
  return out
}

/**
 * What this edge may point at, as a path relative to the editing node.
 *
 * Three tiers, most specific first. The fallback is *everything*: an editor that offered
 * nothing because the ontology was silent would be worse than one that offered too much,
 * and the ontology is silent for 17 of the pairs a real corpus uses.
 */
export function targetCandidates(graph: GraphReport, fromId: string, relationship: string): Candidate[] {
  const from = nodeById(graph, fromId)
  const cls = from?.class ?? ''
  // Every declaration, not the first. `maneuver -[operates-on]->` is declared against three
  // classes, and taking one of them would silently hide two thirds of the legal targets —
  // which reads to a user as "those nodes do not exist".
  const declared = declaredOut(graph, cls, relationship)

  const candidates = new Map<string, Candidate>()
  const offer = (id: string, detail: string, documentation: string, rank: number) => {
    if (id === fromId) return
    const existing = candidates.get(id)
    if (existing && existing.rank <= rank) return
    candidates.set(id, { label: relativeFrom(fromId, id), detail, documentation, rank })
  }

  for (const edge of declared) {
    if (!edge.target) continue
    for (const n of graph.nodes.filter((n) => n.class === edge.target)) {
      offer(n.node, `${n.label || n.node} · ${edge.target}`, edge.description, 0)
    }
  }

  // What the corpus already does with this relationship. Catches `instance-of`, which every
  // node carries, no ontology declares, and which points at an `.ont.yml` rather than at an
  // instance — so no class filter could ever have found it.
  for (const n of graph.nodes) {
    if (n.class !== cls) continue
    for (const l of n.links) {
      if (l.relationship !== relationship || !l.resolved) continue
      const target = nodeById(graph, l.resolved)
      if (target) {
        for (const sibling of graph.nodes.filter((s) => s.class === target.class)) {
          offer(sibling.node, `${sibling.label || sibling.node} · ${target.class}`, '', 1)
        }
      } else {
        offer(l.resolved, 'in use', `Targeted by ${n.node} under the same relationship.`, 1)
      }
    }
  }

  if (candidates.size === 0 && declared.length === 0) {
    for (const n of graph.nodes) {
      offer(n.node, `${n.label || n.node} · ${n.class}`, '', 2)
    }
  }
  return [...candidates.values()]
}

// ── new node ─────────────────────────────────────────────────────────────────

export interface NewNode {
  class: string
  /** Filename stem, kebab-case. */
  name: string
  label: string
  description: string
  relationship: string
  /** Corpus-relative id of the target. */
  target: string
}

/** `Tailwater regime` → `tailwater-regime`. Filenames are stable identity; keep them plain. */
export function slugify(text: string): string {
  return text
    .toLowerCase()
    .normalize('NFKD')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

/**
 * The file, scaffolded from the class's declared properties.
 *
 * Written only once a link has been chosen, which is why `relationship` and `target` are
 * required fields of the input rather than optional ones: a node with no outgoing edge is a
 * lint error the moment it exists, and offering to create one is offering to break the gate.
 */
export function scaffold(input: NewNode, cls: GraphClass | undefined): string {
  const id = `${input.class}/${input.name}.yml`
  const lines = [
    `class: ${input.class}`,
    `label: ${input.label}`,
    `description: ${JSON.stringify(input.description)}`,
  ]
  const props = cls?.properties ?? []
  if (props.length > 0) {
    lines.push('properties:')
    for (const p of props) {
      lines.push(`  # ${p.description}`)
      lines.push(`  ${p.name}: ""   # ${p.type}`)
    }
  }
  lines.push('links:')
  lines.push(`  - target: ${relativeFrom(id, input.target)}`)
  lines.push(`    relationship: ${input.relationship}`)
  return `${lines.join('\n')}\n`
}

File: yidam/cli/src/cmd/graph.rs
//! `yidam graph` — the corpus as nodes, edges, and the classes that license them.
//!
//! Every other report answers a question about the corpus. This one hands over the corpus
//! itself, because the questions an *editor* asks are not a fixed set: where does this edge
//! go, what points back here, what may this class link to, what properties does it carry.
//!
//! # Why the CLI resolves the edges
//!
//! An edge is a filesystem-relative path inside YAML, resolved as `dir.join(target)` against
//! the instance's own directory and normalized so `../class/x.yml` and `class/x.yml` compare
//! equal. That rule is `dangling_edge`'s and `orphan_in`'s, and it is the whole of what makes
//! the graph a graph. A consumer resolving edges itself would be re-deriving it — and would
//! disagree with the gate about which edges are broken, silently, in the direction of
//! "looks fine here".
//!
//! So `resolved` and `exists` are computed here, by the same two lines the checks use.
//!
//! # And why the classes come along
//!
//! `.ont.yml` declares which relationships a class may enter into and which class each one
//! points at. Nothing exposed that outside the file, so an editor offering completion would
//! have had to parse the ontology — a second reader of a shape the prelude defines and
//! `yidam schema` already describes.

use anyhow::Result;
use std::path::{Path, PathBuf};

use crate::parse::CorpusInstance;
use crate::paths::{repo_root, yidam_corpus_dir};
use crate::walk::{walk_corpus_instances, walk_ont_files};

/// A class definition as the ontology writes it.
///
/// Local to this module rather than in `parse.rs`: `CorpusInstance` is read by six other
/// commands and this shape by none of them.
#[derive(Default, serde::Deserialize)]
struct OntologyFile {
    class: Option<String>,
    label: Option<String>,
    description: Option<String>,
    #[serde(default)]
    properties: Vec<OntProperty>,
    #[serde(default)]
    edges: Vec<OntEdge>,
}

#[derive(Clone, serde::Deserialize, serde::Serialize)]
pub struct OntProperty {
    pub name: String,
    #[serde(default)]
    pub r#type: String,
    #[serde(default)]
    pub description: String,
}

/// One relationship a class may enter into.
///
/// `target` is a **class name**, not a path — `holds` on `person` targets `tenure`, and the
/// instances it may point at are the ones in `tenure/`.
#[derive(Clone, serde::Deserialize, serde::Serialize)]
pub struct OntEdge {
    pub relationship: String,
    #[serde(default)]
    pub target: String,
    /// `out` when instances of this class author the link, `in` when the other side does.
    #[serde(default)]
    pub direction: String,
    #[serde(default)]
    pub description: String,
}

#[derive(serde::Serialize)]
pub struct GraphClass {
    pub class: String,
    pub label: String,
    pub description: String,
    pub properties: Vec<OntProperty>,
    pub edges: Vec<OntEdge>,
}

#[derive(serde::Serialize)]
pub struct GraphLink {
    /// Exactly as written, so a consumer can find it in the file.
    pub target: String,
    pub relationship: String,
    /// Corpus-relative path of the resolved target, or empty when it lands outside the
    /// corpus. Resolution is `normalize(dir.join(target))` — the rule `dangling_edge` and
    /// `orphan_in` apply.
    pub resolved: String,
    /// Whether that path is a file. This is the gate's own test, and a consumer MUST NOT
    /// recompute it: an editor that decided for itself would disagree with `lint` about
    /// which edges are broken.
    pub exists: bool,
}

#[derive(serde::Serialize)]
pub struct GraphNode {
    /// Corpus-relative, as `corpus-index` reports it.
    pub node: String,
    pub class: String,
    pub label: String,
    pub description: String,
    pub links: Vec<GraphLink>,
}

#[derive(serde::Serialize)]
pub struct GraphReport {
    /// Repository-relative corpus root, so a consumer building a path does not have to
    /// know that it is `.yidam/corpus`.
    pub corpus_dir: String,
    pub nodes: Vec<GraphNode>,
    pub classes: Vec<GraphClass>,
}

/// Resolve `.` and `..` without touching the filesystem.
///
/// The same walk `lint::checks::normalize` performs. Duplicated rather than shared because
/// that one is private to a module whose surface is checks; if a third caller appears, move
/// it rather than copying it again.
fn normalize(p: &Path) -> PathBuf {
    let mut out = PathBuf::new();
    for c in p.components() {
        match c {
            std::path::Component::ParentDir => {
                out.pop();
            }
            std::path::Component::CurDir => {}
            other => out.push(other),
        }
    }
    out
}

fn slash(p: &Path) -> String {
    p.to_string_lossy().replace('\\', "/")
}

pub(crate) fn graph_data(root: &Path, corpus: &Path) -> GraphReport {
    let nodes = walk_corpus_instances(corpus)
        .iter()
        .map(|path| {
            let text = std::fs::read_to_string(path).unwrap_or_default();
            let inst: CorpusInstance = serde_yaml::from_str(&text).unwrap_or_default();
            let dir = path.parent().unwrap_or(path);
            let links = inst
                .links
                .unwrap_or_default()
                .into_iter()
                .map(|l| {
                    let target = l.target.unwrap_or_default();
                    let absolute = normalize(&dir.join(&target));
                    GraphLink {
                        exists: !target.is_empty() && dir.join(&target).is_file(),
                        resolved: absolute.strip_prefix(corpus).map(slash).unwrap_or_default(),
                        target,
                        relationship: l.relationship.unwrap_or_default(),
                    }
                })
                .collect();
            GraphNode {
                node: slash(path.strip_prefix(corpus).unwrap_or(path)),
                class: inst.class.unwrap_or_default(),
                label: inst.label.unwrap_or_default(),
                description: inst.description.unwrap_or_default(),
                links,
            }
        })
        .collect();

    let classes = walk_ont_files(corpus)
        .iter()
        .map(|path| {
            let text = std::fs::read_to_string(path).unwrap_or_default();
            let ont: OntologyFile = serde_yaml::from_str(&text).unwrap_or_default();
            // The filename is the class when the field is absent: `<class>.ont.yml` is what
            // `graph-check` matches an instance's directory against, so it is the name that
            // actually governs.
            let from_name = path
                .file_name()
                .and_then(|n| n.to_str())
                .and_then(|n| n.strip_suffix(".ont.yml"))
                .unwrap_or_default()
                .to_string();
            GraphClass {
                class: ont.class.unwrap_or(from_name),
                label: ont.label.unwrap_or_default(),
                description: ont.description.unwrap_or_default(),
                properties: ont.properties,
                edges: ont.edges,
            }
        })
        .collect();

    GraphReport {
        corpus_dir: slash(corpus.strip_prefix(root).unwrap_or(corpus)),
        nodes,
        classes,
    }
}

pub(crate) fn render_graph(r: &GraphReport) -> String {
    if r.nodes.is_empty() && r.classes.is_empty() {
        return "Empty corpus — no instances and no class definitions.".to_string();
    }
    let edges: usize = r.nodes.iter().map(|n| n.links.len()).sum();
    let broken: usize = r
        .nodes
        .iter()
        .map(|n| n.links.iter().filter(|l| !l.exists).count())
        .sum();
    let mut out = format!(
        "{} node(s), {} class(es), {edges} edge(s){}\n",
        r.nodes.len(),
        r.classes.len(),
        if broken > 0 {
            format!(" — {broken} pointing at nothing")
        } else {
            String::new()
        }
    );
    for n in &r.nodes {
        for l in &n.links {
            out.push_str(&format!(
                "  {} -[{}]-> {}{}\n",
                n.node,
                l.relationship,
                if l.resolved.is_empty() {
                    &l.target
                } else {
                    &l.resolved
                },
                if l.exists { "" } else { "  (missing)" }
            ));
        }
    }
    out.trim_end().to_string()
}

/// Report the corpus graph: nodes, resolved edges, and the classes that license them.
pub fn graph(format: crate::report::Format) -> Result<()> {
    let root = repo_root()?;
    let data = graph_data(&root, &yidam_corpus_dir(&root));
    if format.is_json() {
        return crate::report::emit(&root, data);
    }
    println!("{}", render_graph(&data));
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    fn corpus() -> (TempDir, PathBuf, PathBuf) {
        let tmp = TempDir::new().unwrap();
        let root = tmp.path().to_path_buf();
        let corpus = root.join(".yidam/corpus");
        std::fs::create_dir_all(corpus.join("concept")).unwrap();
        std::fs::create_dir_all(corpus.join("gauge")).unwrap();
        (tmp, root, corpus)
    }

    fn write(path: &Path, text: &str) {
        std::fs::write(path, text).unwrap();
    }

    /// `../class/x.yml` and `class/x.yml` resolve to the same node, which is the rule
    /// `orphan_in` normalizes for and the reason a consumer must not resolve edges itself.
    #[test]
    fn an_edge_resolves_the_way_the_gate_resolves_it() {
        let (_t, root, corpus) = corpus();
        write(
            &corpus.join("concept/a.yml"),
            "class: concept\nlabel: A\nlinks:\n  - target: ../gauge/g.yml\n    relationship: reads\n",
        );
        write(&corpus.join("gauge/g.yml"), "class: gauge\nlabel: G\n");

        let r = graph_data(&root, &corpus);
        let a = r.nodes.iter().find(|n| n.node == "concept/a.yml").unwrap();
        assert_eq!(a.links[0].resolved, "gauge/g.yml");
        assert!(a.links[0].exists);
        // The raw text is kept, because that is what a consumer has to find in the file.
        assert_eq!(a.links[0].target, "../gauge/g.yml");
    }

    #[test]
    fn an_edge_pointing_at_nothing_is_reported_as_such_and_still_resolves() {
        let (_t, root, corpus) = corpus();
        write(
            &corpus.join("concept/a.yml"),
            "class: concept\nlabel: A\nlinks:\n  - target: ../gauge/gone.yml\n    relationship: reads\n",
        );
        let r = graph_data(&root, &corpus);
        let link = &r.nodes[0].links[0];
        assert!(!link.exists);
        // Resolved anyway: a consumer offering "create this node" needs to know where it
        // would go, and a blank would make a broken edge indistinguishable from an
        // unparseable one.
        assert_eq!(link.resolved, "gauge/gone.yml");
    }

    /// A directory is not a node. `dangling_edge` tests `.exists()`; this tests `is_file()`,
    /// which is the same answer for every real target and the right one for `../gauge`.
    #[test]
    fn a_directory_is_not_a_target() {
        let (_t, root, corpus) = corpus();
        write(
            &corpus.join("concept/a.yml"),
            "class: concept\nlabel: A\nlinks:\n  - target: ../gauge\n    relationship: reads\n",
        );
        assert!(!graph_data(&root, &corpus).nodes[0].links[0].exists);
    }

    #[test]
    fn a_class_carries_its_properties_and_edges() {
        let (_t, root, corpus) = corpus();
        write(
            &corpus.join("concept.ont.yml"),
            "class: concept\nlabel: Concept\ndescription: A unit.\n\
             properties:\n  - name: datum\n    type: string\n    description: Which datum.\n\
             edges:\n  - relationship: reads\n    target: gauge\n    direction: out\n    \
             description: The concept reads a gauge.\n",
        );
        let r = graph_data(&root, &corpus);
        let c = &r.classes[0];
        assert_eq!(c.class, "concept");
        assert_eq!(c.properties[0].name, "datum");
        assert_eq!(c.edges[0].relationship, "reads");
        assert_eq!(c.edges[0].target, "gauge");
        assert_eq!(c.edges[0].direction, "out");
    }

    /// The filename governs when the field is absent — `graph-check` matches an instance's
    /// directory against `<class>.ont.yml`, not against the `class:` field.
    #[test]
    fn a_class_without_a_class_field_is_named_by_its_file() {
        let (_t, root, corpus) = corpus();
        write(&corpus.join("gauge.ont.yml"), "label: Gauge\n");
        assert_eq!(graph_data(&root, &corpus).classes[0].class, "gauge");
    }

    #[test]
    fn the_corpus_root_is_reported_so_nobody_has_to_hardcode_it() {
        let (_t, root, corpus) = corpus();
        assert_eq!(graph_data(&root, &corpus).corpus_dir, ".yidam/corpus");
    }

    #[test]
    fn an_empty_corpus_says_so() {
        let (_t, root, corpus) = corpus();
        assert!(render_graph(&graph_data(&root, &corpus)).contains("Empty corpus"));
    }
}
